infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactoryTest.java

```diff
@@ -36,12 +56,174 @@
 @Listeners(MockitoTestNGListener.class)
 public class KubernetesNamespaceFactoryTest {
   @Mock private KubernetesClientFactory clientFactory;
+
+  @Mock private KubernetesClient k8sClient;
+
+  @Mock
+  private NonNamespaceOperation<
+          Namespace, NamespaceList, DoneableNamespace, Resource<Namespace, DoneableNamespace>>
+      namespaceOperation;
+
   private KubernetesNamespaceFactory namespaceFactory;
 
+  @BeforeMethod
+  public void setUp() throws Exception {
+    lenient().when(clientFactory.create()).thenReturn(k8sClient);
+    lenient().when(k8sClient.namespaces()).thenReturn(namespaceOperation);
+  }
+
+  @Test(
+      expectedExceptions = ConfigurationException.class,
+      expectedExceptionsMessageRegExp =
+          "che.infra.kubernetes.namespace.default or "
+              + "che.infra.kubernetes.namespace.allow_user_defined must be configured")
+  public void
+      shouldThrowExceptionIfNoDefaultNamespaceIsConfiguredAndUserDefinedNamespacesAreNotAllowed()
+          throws Exception {
+    namespaceFactory =
+        new KubernetesNamespaceFactory("predefined", "", "", null, false, clientFactory);
+  }
+
+  @Test
+  public void shouldReturnDefaultNamespaceOnlyWhenItExistsAndUserDefinedIsNotAllowed()
+      throws Exception {
+    prepareNamespaceToBeFoundByName(
+        "che-default",
+        new NamespaceBuilder()
+            .withNewMetadata()
+            .withName("che-default")
+            .endMetadata()
+            .withNewStatus("Active")
+            .build());
+    namespaceFactory =
+        new KubernetesNamespaceFactory("predefined", "", "", "che-default", false, clientFactory);
+
+    List<KubernetesNamespaceMeta> availableNamespaces = namespaceFactory.list();
+    assertEquals(availableNamespaces.size(), 1);
+    KubernetesNamespaceMeta defaultNamespace = availableNamespaces.get(0);
+    assertEquals(defaultNamespace.getName(), "che-default");
+    assertEquals(defaultNamespace.getAttributes().get(DEFAULT_ATTRIBUTE), "true");
+    assertEquals(defaultNamespace.getAttributes().get(PHASE_ATTRIBUTE), "Active");
+  }
+
+  @Test
+  public void shouldReturnDefaultNamespaceOnlyWhenItDoesNotExistAndUserDefinedIsNotAllowed()
```

Comment:
The name of this test (`shouldReturnDefaultNamespaceOnlyWhenItDoesNotExistAndUserDefinedIsNotAllowed`) and the name of the previous test (`shouldReturnDefaultNamespaceOnlyWhenItExistsAndUserDefinedIsNotAllowed`) contradict themselves. What is the correct behavior?

Reply:
The difference it the following shouldReturnDefaultNamespaceOnlyWhen**ItExists** -> shouldReturnDefaultNamespaceOnlyWhen**ItDoesNotExist**

Do you think it makes sense to rephrase the `ItDoesNotExist` one in the way like `shouldReturnDefaultNamespaceHolderOnlyWhenDefaultNamespaceDoesNotExist`?

Reply:
Well, I find the names contradictory because `should return <a> only when <b>` and `should return <a> only when <!b>` just doesn't make sense - if `<a>` should be the right answer for both `<b>` and `<!b>` then there should be just 1 test called `should return <a>`, IMHO.

Reply:
or alternatively, "only when" could be replaced with just "when" in both test names. "Only when" is XOR in my mind :smile:

Reply:
only when replaced with when